### ISSUE TITLE
Out of memory at the beginning of a new epoch after saving checkpoint

### ISSUE THREAD

**Issue by gongbudaizhe** [facebookresearch/fairscale#790]
## ❓ Questions and Help
During the 1st epoch, the training process takes 29G memory.
After saving the 1st checkpoint, the training process takes 31G memory.
At the beginning of the 2nd epoch, the training process crashes with out-of-memory error.

What should be done to solve this problem?

**Comment by blefaudeux** [comment]
> ## ❓ Questions and Help
> 
> During the 1st epoch, the training process takes 29G memory.
> After saving the 1st checkpoint, the training process takes 31G memory.
> At the beginning of the 2nd epoch, the training process crashes with out-of-memory error.
> 
> What should be done to solve this problem?

I think that we would need a little more context, what trainer are you using for instance ? 

For OSS/ShardedDDP it's a very good question, after you checkpoint one of the ranks has a consolidated state, which takes space in memory. You can clear the consolidated state by doing `optimizer._all_states.clear()` for now, but I'll add an API in a minute, it's a good call. This will give you some memory back.

Note that if you're this tight in terms of memory, FSDP will free some more for you, so it could also be a good call to switch to that ?

**Comment by blefaudeux** [comment]
> > ## ❓ Questions and Help
> > During the 1st epoch, the training process takes 29G memory.
> > After saving the 1st checkpoint, the training process takes 31G memory.
> > At the beginning of the 2nd epoch, the training process crashes with out-of-memory error.
> > What should be done to solve this problem?
> 
> I think that we would need a little more context, what trainer are you using for instance ?
> 
> For OSS/ShardedDDP it's a very good question, after you checkpoint one of the ranks has a consolidated state, which takes space in memory. You can clear the consolidated state by doing `optimizer._all_states.clear()` for now, but I'll add an API in a minute, it's a good call. This will give you some memory back.


ah wait, it's been a while.. The state is consolidated in CPU space, so it should not be the issue here

**Comment by blefaudeux** [comment]
I'm still curious about your setup, could you tell us a little more ? Could it also be that the checkpoint is somehow moved to GPU at some point ?

**Comment by min-xu-ai** [comment]
Agreed with @blefaudeux. Can you provide more details on how are checkpoints saved?

**Comment by gongbudaizhe** [comment]
Below is my  script to save the state dict of model, optimizer and scaler:

```python
optimizer.consolidate_state_dict()

# skip saving the queue
state_dict = {}
model_state_dict = model.state_dict()
for key in model_state_dict:
    # we don't need to save the queue
    if 'queue' not in key:
        state_dict[key] = model_state_dict[key]

checkpoint_dict = {
                'epoch': epoch + 1,
                'arch': args.arch,
                'state_dict': state_dict,
                'optimizer': optimizer.state_dict(),
                'scaler': loss_scaler.state_dict(),
                'args': args,
            }

torch.save(checkpoint_dict, save_path)
```

Since OSS already moves params to CPU, I guess it is related to the calls of  model.state_dict() or loss_scaler.state_dict(). Is it possible to move params to CPU in model.state_dict() or loss_scalar.state_dict() ?

**Comment by blefaudeux** [comment]
hmm, checking again the [code in OSS](https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/oss.py#L300) for the state consolidation, it's not obvious that we're keeping a handle on unused tensors. Couple of questions to try to unravel the mystery:
- Would you have an idea of the memory consumption on all ranks ? Is it the same, or is that an issue with the main rank (0 probably) only ? 
- What is your loss scaler ? Is that [ShardedGradScaler](https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/grad_scaler.py#L24) ?
- Anything that you would do at the beginning of an epoch on the dataloading side and which could leak memory ? If you just cycle through epochs on the dataloading (not doing anything with the data), there's no build up ?

**Comment by gongbudaizhe** [comment]
> * Would you have an idea of the memory consumption on all ranks ? Is it the same, or is that an issue with the main rank (0 probably) only ?

All GPUs reaches 31G memory consumption after epoch 1. 

> * What is your loss scaler ? Is that [ShardedGradScaler](https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/grad_scaler.py#L24) ?

Yes

> * Anything that you would do at the beginning of an epoch on the dataloading side and which could leak memory ? If you just cycle through epochs on the dataloading (not doing anything with the data), there's no build up ?

Dataloading won't leak memory is verified. 

Without model.state_dict() and scaler.state_dict(), using optimizer.consolidate_state_dict()  alone increases the memory comsumption to 31G. Nevertheless, the training process won't throw OOM error anymore by removing model.state_dict() and scalar.state_dict().

**Comment by blefaudeux** [comment]
oh wow, great answers, lots of signal there, thank you Bi ! 

Throwing ideas: if there's something wrong in the garbage collection somehow, could you try to `del state_dict` and `del model_dict` after the dump to a file ? 

Checking out the code again, we're deferring to the root optimizer .state_dict() function for some of the state, there are some tensor copies there, and it's true of all ranks. If the state_dict object disappears they should be released, but looks like this fails somehow

**Comment by gongbudaizhe** [comment]
By deleting model_dict of the model and scalar, the training process doesn't throw OOM anymore. However, the training process still occupies 31G memory after calling consolidate_state_dict and the memory is not released. I have tried to delete optimizer._all_states, but it doesn't help.

**Comment by blefaudeux** [comment]
If it doesn't OOM then that's probably what you want, the occupied memory can be a little misleading since it includes a lot of caching which can be released on demand (unless I misunderstood and you're already aware of that, more info [here](https://pytorch.org/docs/stable/notes/cuda.html)). There's a command to manually clear the cache if you want (torch.cuda.empty_cache() I think), or measure the actual used memory at some point in time, if this is an issue then dropping a breakdown of what holding memory would be valuable.

**Comment by blefaudeux** [comment]
by the way thanks for your test of deleting _all_states, it confirms that it's not the issue (these should be cpu memory, but still good to rule that out).

**Comment by gongbudaizhe** [comment]
Thanks, the current state works for me : )
